File: apps/api/src/capabilities/skill-parser.ts
import { z } from 'zod'
import { parse as parseYaml } from 'yaml'
import type {
  SkillDefinition,
  SkillInput,
  CapabilityDefinition,
  ConfigFieldDefinition,
  ToolDefinition,
  ParsedSkillDocument,
} from './types.js'
import type { Prisma } from '@prisma/client'

const ToolDefinitionSchema = z.object({
  name: z.string(),
  description: z.string(),
  prefix: z.string().optional(),
  script: z.string().optional(),
  parameters: z.object({
    type: z.literal('object'),
    properties: z.record(z.unknown()),
    required: z.array(z.string()).optional(),
  }),
})

const SkillDefinitionSchema = z.object({
  name: z.string(),
  slug: z.string().regex(/^[a-z0-9-]+$/, 'Slug must be lowercase alphanumeric with hyphens'),
  description: z.string(),
  version: z.string().default('1.0.0'),
  icon: z.string().optional(),
  category: z.string().default('general'),
  type: z.enum(['bash', 'python', 'js']),
  networkAccess: z.boolean().default(false),
  instructions: z.string(),
  installation: z.string().optional(),
  tools: z.array(ToolDefinitionSchema).min(1),
  inputs: z
    .record(
      z.union([
        z.enum(['var', 'secret', 'textarea']),
        z.object({
          type: z.enum(['var', 'secret', 'textarea']),
          default: z.string().optional(),
          description: z.string().optional(),
          placeholder: z.string().optional(),
        }),
      ]),
    )
    .optional(),
})

const OpenCodeSkillNameSchema = z
  .string()
  .min(1)
  .max(64)
  .regex(/^[a-z0-9]+(-[a-z0-9]+)*$/, 'Skill name must match the OpenCode naming rules')

const ClawbuddySkillSchema = z.object({
  displayName: z.string().optional(),
  version: z.string().default('1.0.0'),
  icon: z.string().optional(),
  category: z.string().default('general'),
  type: z.enum(['bash', 'python', 'js']).optional(),
  tag: z.enum(['bash', 'python', 'js']).optional(),
  networkAccess: z.boolean().default(false),
  installation: z.string().optional(),
  tools: z.array(ToolDefinitionSchema).min(1),
  inputs: z
    .record(
      z.union([
        z.enum(['var', 'secret', 'textarea']),
        z.object({
          type: z.enum(['var', 'secret', 'textarea']),
          default: z.string().optional(),
          description: z.string().optional(),
          placeholder: z.string().optional(),
        }),
      ]),
    )
    .optional(),
})

const OpenCodeFrontmatterSchema = z.object({
  name: OpenCodeSkillNameSchema,
  description: z.string().min(1).max(1024),
  license: z.string().optional(),
  compatibility: z.string().optional(),
  metadata: z.record(z.string()).optional(),
  clawbuddy: ClawbuddySkillSchema,
})

/**
 * Humanize a snake_case or camelCase key into a label.
 * e.g. "aws_access_key_id" -> "Aws Access Key Id"
 */
function humanizeKey(key: string): string {
  return key
    .replace(/[_-]/g, ' ')
    .replace(/([a-z])([A-Z])/g, '$1 $2')
    .replace(/\b\w/g, (c) => c.toUpperCase())
}

/**
 * Convert skill inputs to ConfigFieldDefinition array.
 * Supports both short form ("var"/"secret") and object form ({ type, default, ... }).
 */
function inputsToConfigSchema(inputs: Record<string, SkillInput>): ConfigFieldDefinition[] {
  return Object.entries(inputs).map(([key, input]) => {
    const isObject = typeof input === 'object'
    const inputType = isObject ? input.type : input

    return {
      key,
      label: humanizeKey(key),
      type:
        inputType === 'secret'
          ? ('password' as const)
          : inputType === 'textarea'
            ? ('textarea' as const)
            : ('string' as const),
      required: false,
      envVar: key.toUpperCase(),
      default: isObject ? input.default : undefined,
      description: isObject ? input.description : undefined,
      placeholder: isObject ? input.placeholder : undefined,
    }
  })
}

function buildParsedSkillDocument(
  skill: SkillDefinition,
  format: ParsedSkillDocument['format'],
): ParsedSkillDocument {
  const configSchema = skill.inputs ? inputsToConfigSchema(skill.inputs) : undefined

  const capability: CapabilityDefinition = {
    slug: skill.slug,
    name: skill.name,
    description: skill.description,
    icon: skill.icon,
    category: skill.category ?? 'general',
    version: skill.version,
    tools: skill.tools as ToolDefinition[],
    systemPrompt: skill.instructions,
    configSchema,
    sandbox: {
      networkAccess: skill.networkAccess ?? false,
    },
  }

  const dbData = {
    slug: skill.slug,
    name: skill.name,
    description: skill.description,
    icon: skill.icon,
    category: skill.category ?? 'general',
    version: skill.version,
    toolDefinitions: JSON.parse(JSON.stringify(skill.tools)) as Prisma.InputJsonValue,
    systemPrompt: skill.instructions,
    dockerImage: null,
    packages: [] as string[],
    networkAccess: skill.networkAccess ?? false,
    configSchema: configSchema
      ? (JSON.parse(JSON.stringify(configSchema)) as Prisma.InputJsonValue)
      : undefined,
    builtin: false,
    skillType: skill.type,
    installationScript: skill.installation ?? null,
    source: 'skill' as const,
  }

  return {
    skill,
    capability,
    dbData,
    format,
    storageExtension: format === 'markdown' ? '.md' : '.skill',
    contentType: format === 'markdown' ? 'text/markdown' : 'application/json',
  }
}

function parseMarkdownFrontmatter(content: string): { frontmatter: string; body: string } {
  const normalized = content.replace(/^\uFEFF/, '')
  const match = normalized.match(/^---\r?\n([\s\S]*?)\r?\n---\r?\n?([\s\S]*)$/)

  if (!match) {
    throw new Error('Markdown skills must start with YAML frontmatter delimited by ---')
  }

  return {
    frontmatter: match[1],
    body: match[2].trim(),
  }
}

function parseMarkdownSkill(content: string): ParsedSkillDocument {
  const { frontmatter, body } = parseMarkdownFrontmatter(content)

  let parsedFrontmatter: unknown
  try {
    parsedFrontmatter = parseYaml(frontmatter)
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error)
    throw new Error(`Invalid YAML frontmatter: ${message}`)
  }

  const skillDoc = OpenCodeFrontmatterSchema.parse(parsedFrontmatter)
  const skillType = skillDoc.clawbuddy.type ?? skillDoc.clawbuddy.tag

  if (!skillType) {
    throw new Error('clawbuddy.type is required')
  }

  const skill: SkillDefinition = {
    name: skillDoc.clawbuddy.displayName ?? humanizeKey(skillDoc.name).replace(/\bCli\b/g, 'CLI'),
    slug: skillDoc.name,
    description: skillDoc.description,
    version: skillDoc.clawbuddy.version,
    icon: skillDoc.clawbuddy.icon,
    category: skillDoc.clawbuddy.category,
    type: skillType,
    networkAccess: skillDoc.clawbuddy.networkAccess,
    instructions: body,
    installation: skillDoc.clawbuddy.installation,
    tools: skillDoc.clawbuddy.tools as ToolDefinition[],
    inputs: skillDoc.clawbuddy.inputs,
  }

  return buildParsedSkillDocument(skill, 'markdown')
}

/**
 * Parse and validate a legacy .skill JSON file into a CapabilityDefinition
 * and Prisma-compatible data for upserting.
 */
export function parseSkillFile(raw: unknown): ParsedSkillDocument {
  const skill = SkillDefinitionSchema.parse(raw)

  return buildParsedSkillDocument(skill, 'json')
}

export function parseSkillSource(content: string): ParsedSkillDocument {
  const trimmed = content.trimStart()

  if (trimmed.startsWith('{')) {
    let raw: unknown
    try {
      raw = JSON.parse(content)
    } catch {
      throw new Error('Invalid JSON in legacy .skill file')
    }

    return parseSkillFile(raw)
  }

  return parseMarkdownSkill(content)
}

File: backend/src/clawbuddy/capabilities/skill_parser.py
"""Skill file parser — parses .skill JSON files into capability definitions.

Replaces: apps/api/src/capabilities/skill-parser.ts
"""

from __future__ import annotations

import re
from typing import Any

from pydantic import BaseModel, Field, field_validator

from clawbuddy.capabilities.types import InputType, SkillType


# ---------------------------------------------------------------------------
# Validation schemas (replacing Zod)
# ---------------------------------------------------------------------------

class _ToolParams(BaseModel):
    type: str = "object"
    properties: dict[str, Any] = Field(default_factory=dict)
    required: list[str] | None = None


class _ToolDef(BaseModel):
    name: str
    description: str
    prefix: str | None = None
    script: str | None = None
    parameters: _ToolParams


class _InputObject(BaseModel):
    type: InputType
    default: str | None = None
    description: str | None = None
    placeholder: str | None = None


class _SkillSchema(BaseModel):
    name: str
    slug: str
    description: str
    version: str = "1.0.0"
    icon: str | None = None
    category: str = "general"
    type: SkillType
    network_access: bool = Field(default=False, alias="networkAccess")
    instructions: str
    installation: str | None = None
    tools: list[_ToolDef] = Field(min_length=1)
    inputs: dict[str, Any] | None = None

    model_config = {"populate_by_name": True}

    @field_validator("slug")
    @classmethod
    def _validate_slug(cls, v: str) -> str:
        if not re.fullmatch(r"[a-z0-9-]+", v):
            raise ValueError("Slug must be lowercase alphanumeric with hyphens")
        return v


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _humanize_key(key: str) -> str:
    """Convert snake_case or camelCase key to a human-readable label.

    e.g. "aws_access_key_id" -> "Aws Access Key Id"
    """
    # Replace underscores with spaces
    result = key.replace("_", " ")
    # Insert space before uppercase letters in camelCase
    result = re.sub(r"([a-z])([A-Z])", r"\1 \2", result)
    # Capitalize each word
    return result.title()


def _inputs_to_config_schema(
    inputs: dict[str, Any],
) -> list[dict[str, Any]]:
    """Convert skill inputs to ConfigFieldDefinition dicts.

    Supports both short form ("var" / "secret" / "textarea") and object form
    ({ type, default, description, placeholder }).
    """
    result: list[dict[str, Any]] = []
    for key, raw_input in inputs.items():
        if isinstance(raw_input, str):
            input_type: str = raw_input
            default = None
            description = None
            placeholder = None
        elif isinstance(raw_input, dict):
            parsed = _InputObject.model_validate(raw_input)
            input_type = parsed.type
            default = parsed.default
            description = parsed.description
            placeholder = parsed.placeholder
        else:
            raise ValueError(f"Invalid input definition for key '{key}'")

        # Map input types to config field types
        if input_type == "secret":
            field_type = "password"
        elif input_type == "textarea":
            field_type = "textarea"
        else:
            field_type = "string"

        field: dict[str, Any] = {
            "key": key,
            "label": _humanize_key(key),
            "type": field_type,
            "required": False,
            "envVar": key.upper(),
        }
        if default is not None:
            field["default"] = default
        if description is not None:
            field["description"] = description
        if placeholder is not None:
            field["placeholder"] = placeholder

        result.append(field)
    return result


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

class ParsedSkill:
    """Result of parsing a .skill file."""

    __slots__ = ("skill", "capability", "db_data")

    def __init__(
        self,
        skill: _SkillSchema,
        capability: dict[str, Any],
        db_data: dict[str, Any],
    ) -> None:
        self.skill = skill
        self.capability = capability
        self.db_data = db_data


def parse_skill_file(raw: Any) -> ParsedSkill:
    """Parse and validate a .skill file JSON into a capability definition.

    Returns a ParsedSkill with:
    - skill: validated SkillDefinition
    - capability: CapabilityDefinition dict
    - db_data: dict suitable for upserting into the Capability DB model
    """
    skill = _SkillSchema.model_validate(raw)

    config_schema = (
        _inputs_to_config_schema(skill.inputs) if skill.inputs else None
    )

    # Build tool definitions as plain dicts
    tools = [
        {
            "name": t.name,
            "description": t.description,
            **({"prefix": t.prefix} if t.prefix else {}),
            **({"script": t.script} if t.script else {}),
            "parameters": t.parameters.model_dump(),
        }
        for t in skill.tools
    ]

    capability: dict[str, Any] = {
        "slug": skill.slug,
        "name": skill.name,
        "description": skill.description,
        "icon": skill.icon,
        "category": skill.category,
        "version": skill.version,
        "tools": tools,
        "systemPrompt": skill.instructions,
        "sandbox": {
            "networkAccess": skill.network_access,
        },
    }
    if config_schema:
        capability["configSchema"] = config_schema

    db_data: dict[str, Any] = {
        "slug": skill.slug,
        "name": skill.name,
        "description": skill.description,
        "icon": skill.icon,
        "category": skill.category,
        "version": skill.version,
        "tool_definitions": tools,
        "system_prompt": skill.instructions,
        "docker_image": None,
        "packages": [],
        "network_access": skill.network_access,
        "config_schema": config_schema,
        "builtin": False,
        "skill_type": skill.type,
        "installation_script": skill.installation,
        "source": "skill",
    }

    return ParsedSkill(skill=skill, capability=capability, db_data=db_data)
